### DISCUSSION TITLE
Can I send network request in custom Audit?

### DISCUSSION THREAD

**Discussion by lamysafari** [GoogleChrome/lighthouse#15367] [Q&A]
For example , I want to check if every Anchor links in document is valid by checking status code. There is a AnchorElements gather, so I can collect links,  but how can I send request? In a custom audit or create a new gather?

**Comment by connorjclark** [comment]
In your custom audit, you can just use typical Node network utilities. This is reaching outside the framework Lighthouse provides, and isn't something we ever do in our own audits, but for a custom audit it's probably fine.

If you want to do it in a gatherer (this makes it possible to serialize results to disk and keeps the auditing portion side-effect free / not reliant on a network), here is some high level guidance:

- Define a new gatherer that has AnchorElements as a dependency - ex: https://github.com/GoogleChrome/lighthouse/blob/e955992ea00e05b95f9c13496b88abee72a67af1/core/gather/gatherers/link-elements.js#L106 (@adamraine that should be possible, right? Gatherer deps aren't just for Trace/DevtoolsLog?)
- Use `driver.fetcher.fetchResource` to fetch a URL https://github.com/GoogleChrome/lighthouse/blob/e955992ea00e05b95f9c13496b88abee72a67af1/core/gather/gatherers/source-maps.js#L27 , which gives you the content and status code
- Use this custom gatherer in your custom audit. An example config that may help: https://github.com/GoogleChrome/lighthouse/blob/e955992ea00e05b95f9c13496b88abee72a67af1/docs/recipes/custom-audit/custom-config.js#L11

**Reply by adamraine** [reply]
> (@adamraine that should be possible, right? Gatherer deps aren't just for Trace/DevtoolsLog?)

Not for anchor elements, we don't define a `symbol` on the anchor elements gatherer. Might be worth adding one for every gatherer or removing this symbol logic completely though.

**Reply by adamraine** [reply]
For reference, the justification for symbol stuff: https://github.com/GoogleChrome/lighthouse/pull/12017#discussion_r567073904

**Reply by connorjclark** [reply]
Thanks Adam. Tracking in #15368
